optuna/integration/dask.py

```diff
@@ -0,0 +1,733 @@
+    from distributed.worker import get_client
+
+# Serialization utilities
+
+
+def _serialize_datetime(dt: datetime.datetime) -> dict:
+    return {"__datetime__": True, "as_str": dt.strftime("%Y%m%dT%H:%M:%S.%f")}
+
+
+def _deserialize_datetime(data: dict) -> datetime.datetime:
+    return datetime.datetime.strptime(data["as_str"], "%Y%m%dT%H:%M:%S.%f")
+
+
+def _serialize_trialstate(state: TrialState) -> str:
+    return state.name
+
+
+def _deserialize_trialstate(state: str) -> TrialState:
+    return getattr(TrialState, state)
+
+
+def _serialize_frozentrial(trial: FrozenTrial) -> dict:
+    data = trial.__dict__.copy()
+    data["state"] = _serialize_trialstate(data["state"])
+    attrs = [a for a in data.keys() if a.startswith("_")]
+    for attr in attrs:
+        data[attr[1:]] = data.pop(attr)
+    data["distributions"] = {k: distribution_to_json(v) for k, v in data["distributions"].items()}
+    if data["datetime_start"] is not None:
+        data["datetime_start"] = _serialize_datetime(data["datetime_start"])
+    if data["datetime_complete"] is not None:
+        data["datetime_complete"] = _serialize_datetime(data["datetime_complete"])
+    data["value"] = None
+    return data
+
+
+def _deserialize_frozentrial(data: dict) -> FrozenTrial:
+    data["state"] = _deserialize_trialstate(data["state"])
+    data["distributions"] = {k: json_to_distribution(v) for k, v in data["distributions"].items()}
+    if data["datetime_start"] is not None:
+        data["datetime_start"] = _deserialize_datetime(data["datetime_start"])
+    if data["datetime_complete"] is not None:
+        data["datetime_complete"] = _deserialize_datetime(data["datetime_complete"])
+    trail = FrozenTrial(**data)
+    return trail
+
+
+def _serialize_studydirection(direction: StudyDirection) -> str:
+    return direction.name
+
+
+def _deserialize_studydirection(data: str) -> StudyDirection:
+    return getattr(StudyDirection, data)
+
+
+class _OptunaSchedulerExtension:
+    def __init__(self, scheduler: "distributed.Scheduler"):
+        self.scheduler = scheduler
+        self.storages: Dict[str, BaseStorage] = {}
+
+        methods = [
+            "create_new_study",
+            "delete_study",
+            "set_study_user_attr",
+            "set_study_system_attr",
+            "set_study_directions",
+            "get_study_id_from_name",
+            "get_study_name_from_id",
+            "get_study_directions",
+            "get_study_user_attrs",
+            "get_study_system_attrs",
+            "get_all_studies",
+            "create_new_trial",
+            "set_trial_param",
+            "get_trial_id_from_study_id_trial_number",
+            "get_trial_number_from_id",
+            "get_trial_param",
+            "set_trial_state_values",
+            "set_trial_intermediate_value",
+            "set_trial_user_attr",
+            "set_trial_system_attr",
+            "get_trial",
+            "get_all_trials",
+            "get_n_trials",
```

Comment:
Is it better to add `get_best_trial`, `get_trial_params`, `get_trial_user_attr`, and `get_trial_system_attrs` here?

Reply:
As is, these methods work since they internally only reference other abstract storage methods which are implemented in this PR. So the existing implementation in `BaseStorage` should be sufficient. Though, you're correct that things might be sub-optimal as we're sending intermediate results back and forth between the client and scheduler processes.

Definitely open to thoughts from others, but my personal preference would be to hold off on adding these methods in this PR. Then, after this PR is merged, we can do some benchmarking to see how much benefit we'd get in practice by adding these derivative methods. If it turns out to be a sizable performance impact, then adding these methods should be relatively straightforward.

Reply:
Sure 👍  In terms of performance impact, it obviously depends on which storage our users use. Probably you’ve already noticed, but RDBStorage overrides these methods to issue more efficient SQL queries.

https://github.com/optuna/optuna/blob/f02f87dd3e1dfd3b9224917b55caec79f091bc2e/optuna/storages/_rdb/storage.py#L402-L422

https://github.com/optuna/optuna/blob/f02f87dd3e1dfd3b9224917b55caec79f091bc2e/optuna/storages/_rdb/storage.py#L951-L967

Anyway, as you said, adding these methods should be relatively straightforward. So you can work on it as a follow-up.